File: leetcode/graph/323_Number_of_Connected_Components_in_an_Undirected_Graph/codes/union.c
#include <stdio.h>
#include <stdlib.h>

int find(int* parent, int x) {
    if (parent[x] != x)
        parent[x] = find(parent, parent[x]);
    return parent[x];
}

void unionSet(int* parent, int x, int y) {
    int px = find(parent, x);
    int py = find(parent, y);
    if (px != py)
        parent[px] = py;
}

int countComponents(int n, int** edges, int edgesSize, int* edgesColSize) {
    int* parent = (int*)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) parent[i] = i;

    for (int i = 0; i < edgesSize; i++) {
        int u = edges[i][0];
        int v = edges[i][1];
        unionSet(parent, u, v);
    }

    int* seen = (int*)calloc(n, sizeof(int));
    int count = 0;
    for (int i = 0; i < n; i++) {
        int root = find(parent, i);
        if (!seen[root]) {
            seen[root] = 1;
            count++;
        }
    }

    free(parent);
    free(seen);
    return count;
}

File: leetcode/graph/323_Number_of_Connected_Components_in_an_Undirected_Graph/codes/union.py
from typing import List

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        parent = list(range(n))

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])  
            return parent[x]

        def union(x, y):
            px, py = find(x), find(y)
            if px != py:
                parent[px] = py

        for u, v in edges:
            union(u, v)

        return len(set(find(i) for i in range(n)))
